src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```diff
@@ -176,4 +193,122 @@ private void serializeModel(TrainingJob trainingJob, ActionListener<IndexRespons
+                    modelDao.update(model, new ActionListener<IndexResponse>() {
+                        @Override
+                        public void onResponse(IndexResponse indexResponse) {
+                            logger.info(
+                                "Model "
+                                    + indexResponse.getId()
+                                    + " updated from "
+                                    + ModelState.TRAINING
+                                    + " to "
+                                    + ModelState.FAILED
+                                    + " due to node drop"
+                            );
+                        }
+
+                        @Override
+                        public void onFailure(Exception e) {
+                            logger.info("Failed to update model after node drop", e);
+                        }
+                    });
+                }
+            }
+        }
+    }
+
+    protected void updateModelsNodesRemoved(List<DiscoveryNode> removedNodes) throws IOException, InterruptedException, ExecutionException {
+        if (modelDao.isCreated()) {
+            List<String> modelIds = searchModelIds();
+            for (DiscoveryNode removedNode : removedNodes) {
+                for (String modelId : modelIds) {
+                    Model model = modelDao.get(modelId);
+                    ModelMetadata modelMetadata = model.getModelMetadata();
+                    if (modelMetadata.getNodeAssignment().equals(removedNode.getEphemeralId())
+                        && modelMetadata.getState().equals(ModelState.TRAINING)) {
+                        modelMetadata.setState(ModelState.ZOMBIE);
+                        modelMetadata.setError("A node dropped and left the model training process in a zombie state");
+                        modelDao.update(model, new ActionListener<IndexResponse>() {
+                            @Override
+                            public void onResponse(IndexResponse indexResponse) {
+                                logger.info(
+                                    "Model "
+                                        + indexResponse.getId()
+                                        + " updated from "
+                                        + ModelState.TRAINING
+                                        + " to "
+                                        + ModelState.ZOMBIE
+                                        + " due to node drop"
+                                );
+                            }
+
+                            @Override
+                            public void onFailure(Exception e) {
+                                logger.info("Failed to update model after node drop", e);
+                            }
+                        });
+                    }
+                }
+            }
+        }
+    }
+
+    private List<String> searchModelIds() throws IOException, InterruptedException {
```

Comment:
Is there a reason why this has to be a sync call?

Reply:
We need the result (the list of modelIds) to update the correct models in the methods where it is called

Reply:
You can pass callback function to update models.